Update the release process guide (#200)
submariner-io/submariner-website: src/content/contributing/release-process/_index.en.md

The Go dependency hierarchy flow is as follows:

`shipyard` <- `admiral` <- `[submariner, lighthouse]` <- `submariner-operator`

Note that the `submariner` and `lighthouse` projects are siblings and thus do not depend on one another. Also the
`submariner-operator` components expect that `lighthouse` and `submariner` are aligned on the same exact version.

The Docker image dependency hierarchy flow is as follows:

`subctl binary` <- `shipyard dapper base image` <- `[admiral, submariner, lighthouse, submariner-operator]`

The Dapper base image that is provided by `shipyard` for building and E2E tests in all of the other projects pulls in the
`subctl` binary.

## Release versions

The `vx.x.x` version format is used for the git projects while `x.x.x` is used for Docker images.

The typical workflow is to first create release candidate(s) for testing before creating the final release. The suggested
naming convention is to append `-rcN` to the final version, for example `v0.5.0-rc0`, `v0.5.0-rc1` etc.

Sometimes you may want to create a specific project release for testing prior to creating a release candidate. In this
case, the suggested naming convention is to append `-preN`.

> ## Create Submariner product release

Reviewer: ```suggestion
## Create a submariner-io github org product release
```

Suggestion to clarify further based on mike comments

Author: This section is meant as an intro to creating a full Submariner **product** release, which encompasses all the github projects. The intro segues into the individual steps as subsections. I could add a subsection for  "Create a GitHub 
 project release" instead of inlining it.

Reviewer: Ah ok, this is ## and the other is ###

Reviewer: ok, my suggestions should be ignored then

weight: 10
---

This section describes how to create a Submariner product release. It is assumed that you are familiar with the various
Submariner projects and their repositories and are familiar with git and GitHub.

## Project dependencies

> The Submariner projects have a dependency hierarchy with respect to their Go libraries and docker images.

Reviewer 2: ```suggestion
The Submariner projects have a dependency hierarchy with respect to their Go libraries and Docker images.
```

https://kubernetes.io/docs/contribute/style/style-guide/#kubernetes-io-word-list

Therefore the releases for each individual project must be created in a specific order.

The Go dependency hierarchy flow is as follows:

`shipyard` <- `admiral` <- `[submariner, lighthouse]` <- `submariner-operator`

Note that the `submariner` and `lighthouse` projects are siblings and thus do not depend on one another. Also the
`submariner-operator` components expect that `lighthouse` and `submariner` are aligned on the same exact version.

> The docker image dependency hierarchy flow is as follows:

Reviewer 2: ```suggestion
The Docker image dependency hierarchy flow is as follows:
```

`subctl binary` <- `shipyard dapper base image` <- `[admiral, submariner, lighthouse, submariner-operator]`

The dapper base image that is provided by `shipyard` for building and E2E tests in all of the other projects pulls in the
`subctl` binary.
 
## Release versions

> The `vx.x.x` version format is used for the git projects while `x.x.x` is used for docker images.

Reviewer 2: ```suggestion
The `vx.x.x` version format is used for the git projects while `x.x.x` is used for Docker images.
```

Sometimes you may want to create a specific project release for testing prior to creating a release candidate. In this
case, the suggested naming convention is to append `-preN`.

## Create Submariner product release

The following outlines the steps to be taken in order to create a full Submariner product release. As an example, we'll
use version `v0.5.0`.

Project releases can either be created via the git CLI or the GitHub UI. This guide uses the GitHub UI as it is
simpler:

1) Navigate to the project's main releases page.

2) Click `Draft a new release`

3) Fill in the `Tag version` field with the new version.

4) Fill in the `Release title` field as appropriate. Typically this is just the version (sans the leading `v`) as a release
   usually contains various changes. But if it's a targeted release then the `Release title` can reflect that.
   
5) Enter information to describe the release. This is optional for pre-releases but should be filled in for
   a final release.
   
6) If this is not a final release, mark the checkbox `This is a pre-release`.

7) Click `Publish release`.

> For most projects a GitHub action job will be initiated to build release artifacts and publish to quay. This will take

Reviewer 2: ```suggestion
For most projects a GitHub action job will be initiated to build release artifacts and publish to Quay. This will take
```

several minutes. You can monitor the progress from the project's main page. A small yellow circle should be present to
the right of the heading above the file listing which indicates it's in progress. You can click it to see details. When
complete, it will change to either a green check mark or a red X.

> ### Step 0: Create a `subctl` pre-release, optional

Reviewer 3: Shouldn't this come before ` ## Create Submariner product release ` since otherwise your release jobs wont be using the latest subctl?

Or perhaps clarify in that section that it's the general way of creating a release, not instructions to actually start doing. I can see how someone will read this guide and just start that section on `Submariner` project before reading the rest..

Author: > Shouldn't this come before `## Create Submariner product release` since otherwise your release jobs wont be using the latest subctl?

I consider it part of creating a Submariner product release, ie it's an initial step that may or may not be need to be done, as indicated by the conditional semantics like "optional", so I don't think it should be before it.  

> 
> Or perhaps clarify in that section that it's the general way of creating a release, not instructions to actually start doing. 

I assume you mean don't include it as a Step. We could include it as part of the previous section but it may get missed if someone skims the previous sections and goes right to the Steps. I prefer it as a Step. We could just make it Step 1 and/or don't treat it as optional since it doesn't hurt to create the pre-release and would be safest to just do it. 

> I can see how someone will read this guide and just start that section on `Submariner` project before reading the rest..

I'm not sure what you mean here.

Author: OK - I get it now - I clarified it in https://github.com/submariner-io/submariner-website/pull/200/commits/bbd59f19485581b5be455acdd0b24d4c236c59da

> Since `subctl` is provided by `shipyard`'s dapper base image, if there are specific changes to `subctl` which other

Reviewer 2: ```suggestion
Since `subctl` is provided by `shipyard`'s Dapper base image, if there are specific changes to `subctl` which other
```

projects require, you will need to create a pre-release (for example `v0.5.0-pre0`) of the `submariner-operator` project: 

> 1) Navigate to the [releases](https://github.com/submariner-io/submariner-operator/releases) page and create the release

Reviewer 2: Might be good to extract this link to a reference, as it's used quite a few times.

Author: I assume you mean that "releases"  is used several times but the underlying URL is different.

Reviewer 2: Oh I missed that they are to different projects' release pages, my bad. Okay, the way you have it is perfect then!

   with the new version.

2) In the `shipyard` project, edit `package/Dockerfile.shipyard-dapper-base` and set `SUBCTL_VERSION` to the new version.

> 3) Commit the change, create a pull request and merge it.

Reviewer 3: ```suggestion
3) Commit the change, create a pull request and get it reviewed and merged.
```

Reviewer 3: You might also want to instruct to label the PR with `test-projects` so it runs E2E on the consuming projects as well.

### Step 1: Create release for the `shipyard` project

Navigate to the [releases](https://github.com/submariner-io/shipyard/releases) page and create the release with the new
version. This will initiate a job to build the dapper base image. Once successfully completed, the generated image

> version (`0.5.0`) should be available on quay here: 

Reviewer 2: ```suggestion
version (`0.5.0`) should be available on Quay here: 
```

> https://quay.io/repository/submariner/shipyard-dapper-base?tab=tags

### Step 2: Create release for the `admiral` project

1) Pin the `shipyard` dapper base image to the new version. Edit `Dockerfile.dapper` and, on the first line, change
   the `shipyard-dapper-base` image version from `devel` to the new version (`0.5.0`).

2) Update the _go.mod_ and _go.sum_ references for `shipyard` to the new version:

```
make shell
go get github.com/submariner-io/shipyard@v0.5.0
go mod vendor
go mod tidy

> exit

Reviewer 3: Maybe this should be in the Makefile, eg `make prep-release` ?

Author: yeah it could be.

```

> 3) Commit the changes, create a pull request and merge it.

Reviewer 3: ```suggestion
3) Commit the changes, create a pull request and get it reviewed and merged.
```

Author: I didn't explicitly mention about getting it reviewed  as I think that's just part of the merging details plus it's problematic to require these PRs to go thru the normal review  process as that could drag out a release for days with waiting for reviewers for each PR.

4) Navigate to [releases](https://github.com/submariner-io/admiral/releases) and create the release for the new version.

### Step 3: Create releases for the `submariner` and `lighthouse` projects

These can be done in any order or in parallel and the process is the same.

> 1) Pin the `shipyard` dapper base image to the new version. Edit _Dockerfile.dapper_ and, on the first line, change

Reviewer 2: ```suggestion
1) Pin the `shipyard` Dapper base image to the new version. Edit _Dockerfile.dapper_ and, on the first line, change
```

   the `shipyard-dapper-base` image version from `devel` to the new version (`0.5.0`).

2) Update the _go.mod_ and _go.sum_ references for the dependent projects to the new version:

```
make shell
go get github.com/submariner-io/shipyard@v0.5.0
go get github.com/submariner-io/admiral@v0.5.0
go mod vendor

> go mod tidy

Reviewer 3: Same here re `make prep-release`

```

> 3) Commit the changes, create a pull request and merge it.

Reviewer 3: ```suggestion
3) Commit the changes, create a pull request and get it reviewed and merged.
```

4) Navigate to the project's releases page and create the release for the new version.

> Wait for the project's new images to be generated and published to quay.

Reviewer 2: ```suggestion
Wait for the project's new images to be generated and published to Quay.
```

For `submariner`:

> https://quay.io/repository/submariner/submariner?tab=tags
> https://quay.io/repository/submariner/submariner-route-agent?tab=tags
> https://quay.io/repository/submariner/submariner-globalnet?tab=tags

For `lighthouse`:

> https://quay.io/repository/submariner/lighthouse-agent?tab=tags
> https://quay.io/repository/submariner/lighthouse-coredns?tab=tags

### Step 4: Create release for the `submariner-operator` project

1) Pin the `shipyard` dapper base image to the new version. Edit `Dockerfile.dapper` and, on the first line, change
   the `shipyard-dapper-base` image version from `devel` to the new version (`0.5.0`).
   
2) Update the _go.mod_ and _go.sum_ references for the dependent projects to the new version:

```
make shell
go get github.com/submariner-io/shipyard@v0.5.0
go get github.com/submariner-io/admiral@v0.5.0
go get github.com/submariner-io/submariner@v0.5.0
go get github.com/submariner-io/lighthouse@v0.5.0
go mod vendor

> go mod tidy

Reviewer 3: Same here re `make prep-release`

```

3) Edit _pkg/versions/versions.go_ and update the *Version constants to the new version:

```
DefaultSubmarinerOperatorVersion = "0.5.0"
DefaultSubmarinerVersion         = "0.5.0"
DefaultLighthouseVersion         = "0.5.0"
```

> 3) Commit the changes, create a pull request and merge it.

Reviewer 3: ```suggestion
3) Commit the changes, create a pull request and get it reviewed and merged.
```

4) Navigate to [releases](https://github.com/submariner-io/submariner-operator/releases) and create the release for the
   new version.

> Once the build job successfully completes, the generated image version (`0.5.0`) should be available on quay here: 

Reviewer 2: ```suggestion
Once the build job successfully completes, the generated image version (`0.5.0`) should be available on Quay here: 
```

> https://quay.io/repository/submariner/submariner-operator?tab=tags

The `subctl` binaries for the various platforms should be listed under the `Assets` section for the new release on the
main [releases](https://github.com/submariner-io/submariner-operator/releases) page.
 
### Step 5: Update `subctl` version in `shipyard`

> Update the dapper base image to pull in the latest `subctl` binary:

Reviewer 2: ```suggestion
Update the Dapper base image to pull in the latest `subctl` binary:
```

1) In the `shipyard` project, edit `package/Dockerfile.shipyard-dapper-base` and set `SUBCTL_VERSION` to the new version.

> 2) Commit the change, create a pull request and merge it.

Reviewer 3: ```suggestion
2) Commit the change, create a pull request and get it reviewed and merged.
```

### Step 6: Add release notes
 

> If this is a final release, add a section for it on this website's [release notes](../..//releases/) page.

Reviewer 3: ```suggestion
If this is a final release, add a section for it on this website's [release notes](../../releases/) page.
```

1) Clone the [submariner-website](https://github.com/submariner-io/submariner-website) project.

> 2) Open src/content/releases/_index.en.md and make changes.

Reviewer 2: ```suggestion
2) Open `src/content/releases/_index.en.md` and make changes.
```

> 3) Commit the changes and create a pull request.

Reviewer 3: ```suggestion
3) Commit the changes, create a pull request and get it reviewed and merged.
```

weight: 10
---

This section describes how to create a Submariner product release. It is assumed that you are familiar with the various
Submariner projects and their repositories and are familiar with Git and GitHub.

## Project dependencies

The Submariner projects have a dependency hierarchy with respect to their Go libraries and Docker images.
Therefore the releases for each individual project must be created in a specific order.

The Go dependency hierarchy flow is as follows:

`shipyard` <- `admiral` <- `[submariner, lighthouse]` <- `submariner-operator`

Note that the `submariner` and `lighthouse` projects are siblings and thus do not depend on one another. Also the
`submariner-operator` components expect that `lighthouse` and `submariner` are aligned on the same exact version.

The Docker image dependency hierarchy flow is as follows:

`subctl binary` <- `shipyard dapper base image` <- `[admiral, submariner, lighthouse, submariner-operator]`

> The dapper base image that is provided by `shipyard` for building and E2E tests in all of the other projects pulls in the

Reviewer 2: I think "Dapper" should be capitalized in most cases here, where it's not part of some `code literal`.
